## Ques:4. Analyze Wine quality data set using linear regression (download data from UCI web repository)

### (a) Analyze the data with normalization and without normalization.

In [202]:
df6 =pd.read_csv('winequality.csv',header=None,skiprows=1)
df6


0      1     2    3      4     5     6        7     8     9     10  11
0      7.4  0.700  0.00  1.9  0.076  11.0  34.0  0.99780  3.51  0.56   9.4   5
1      7.8  0.880  0.00  2.6  0.098  25.0  67.0  0.99680  3.20  0.68   9.8   5
2      7.8  0.760  0.04  2.3  0.092  15.0  54.0  0.99700  3.26  0.65   9.8   5
3     11.2  0.280  0.56  1.9  0.075  17.0  60.0  0.99800  3.16  0.58   9.8   6
4      7.4  0.700  0.00  1.9  0.076  11.0  34.0  0.99780  3.51  0.56   9.4   5
...    ...    ...   ...  ...    ...   ...   ...      ...   ...   ...   ...  ..
1594   6.2  0.600  0.08  2.0  0.090  32.0  44.0  0.99490  3.45  0.58  10.5   5
1595   5.9  0.550  0.10  2.2  0.062  39.0  51.0  0.99512  3.52  0.76  11.2   6
1596   6.3  0.510  0.13  2.3  0.076  29.0  40.0  0.99574  3.42  0.75  11.0   6
1597   5.9  0.645  0.12  2.0  0.075  32.0  44.0  0.99547  3.57  0.71  10.2   5
1598   6.0  0.310  0.47  3.6  0.067  18.0  42.0  0.99549  3.39  0.66  11.0   6

[1599 rows x 12 columns]

In [226]:
gradient_descent(df6,10000,0.0001)

(0.5407800555554853,
 array([ 0.06718371, -0.31701839, -0.03782658,  0.05227645, -0.22075078,
         0.06064083, -0.18238542, -0.05836576, -0.09727041,  0.30699806,
         0.35491125]),
 0.016671086676918086)

In [229]:
avg_mse=np.mean(mse_list)
print(avg_mse)

0.016905940702166767


In [230]:
np.std(y-y_pred)

0.1291165623302685

In [231]:
df6_normalized=normalisation(df6)
df6_normalized

0         1     2         3         4         5         6   \
0     0.247788  0.397260  0.00  0.068493  0.106845  0.140845  0.098940   
1     0.283186  0.520548  0.00  0.116438  0.143573  0.338028  0.215548   
2     0.283186  0.438356  0.04  0.095890  0.133556  0.197183  0.169611   
3     0.584071  0.109589  0.56  0.068493  0.105175  0.225352  0.190813   
4     0.247788  0.397260  0.00  0.068493  0.106845  0.140845  0.098940   
...        ...       ...   ...       ...       ...       ...       ...   
1594  0.141593  0.328767  0.08  0.075342  0.130217  0.436620  0.134276   
1595  0.115044  0.294521  0.10  0.089041  0.083472  0.535211  0.159011   
1596  0.150442  0.267123  0.13  0.095890  0.106845  0.394366  0.120141   
1597  0.115044  0.359589  0.12  0.075342  0.105175  0.436620  0.134276   
1598  0.123894  0.130137  0.47  0.184932  0.091820  0.239437  0.127208   

            7         8         9         10   11  
0     0.567548  0.606299  0.137725  0.153846  0.4  
1     0.494126  0.362205  0.209581  0.215385  0.4  
2     0.508811  0.409449  0.191617  0.215385  0.4  
3     0.582232  0.330709  0.149701  0.215385  0.6  
4     0.567548  0.606299  0.137725  0.153846  0.4  
...        ...       ...       ...       ...  ...  
1594  0.354626  0.559055  0.149701  0.323077  0.4  
1595  0.370778  0.614173  0.257485  0.430769  0.6  
1596  0.416300  0.535433  0.251497  0.400000  0.6  
1597  0.396476  0.653543  0.227545  0.276923  0.4  
1598  0.397944  0.511811  0.197605  0.400000  0.6  

[1599 rows x 12 columns]

In [232]:
gradient_descent(df6_normalized,10000,0.0001)

(0.5407800555554853,
 array([ 0.06718371, -0.31701839, -0.03782658,  0.05227645, -0.22075078,
         0.06064083, -0.18238542, -0.05836576, -0.09727041,  0.30699806,
         0.35491125]),
 0.016671086676918086)

In [233]:
avg_mse=np.mean(mse_list)
print(avg_mse)

0.016905940702166767


In [234]:
np.std(y-y_pred)

0.1291165623302685

### (b) Describe how you applied normalization techniques on training and testing data.

### From the given Data Set  Found the max , min for each and every column and stored them in variables pmax , pmin
### Now divide the data set into training data and testing data
### Now apply normalization seperately for both traning and testing data with the help of pmax and pmin
def normalisation(df):

   
    for i in df.columns:
        pmax=max(df[i])
        pmin=min(df[i])
        df[i]= np.array((df[i]-pmin)/(pmax-pmin))
   
  
    return df


### (c) Apply k fold cross validation and hold out method.


## K_Fold Cross Validation Technique

In [251]:
KFold_cross_validation_train(df6,10000,0.00000001,5)

 Data Set Lenght : 1599 
1280
1280
w0:0.06094671458116376 w:[0.02065262 0.01579005 0.01789505 0.00692555 0.00721473 0.01276403
 0.00782086 0.02889846 0.02709044 0.01233356 0.02192146] Mse:0.2024814645280697
1280
1280
w0:0.05922424928606754 w:[0.01776958 0.01574372 0.0148557  0.00631152 0.00719641 0.0124768
 0.00809392 0.02668626 0.02738947 0.01173219 0.02003132] Mse:0.19310459562774823
1280
1280
w0:0.0601648480417416 w:[0.02015146 0.01572017 0.01757305 0.00676811 0.00739068 0.0125346
 0.00821757 0.02887484 0.02665666 0.01257963 0.02037238] Mse:0.19693923442069022
1280
1280
w0:0.05821380502616128 w:[0.01989397 0.01578958 0.01656865 0.00668128 0.00724503 0.01190881
 0.00815934 0.0293612  0.02615363 0.01209478 0.01836876] Mse:0.18521939643978919
1280
1280
w0:0.06032254262033572 w:[0.02157634 0.01552417 0.0183055  0.00688993 0.00751954 0.01210799
 0.00816103 0.02992801 0.02646773 0.01252799 0.02031071] Mse:0.1978345358060857


## Hold Out Cross Validation Technique

In [295]:
holdout_method(df6,10000,0.00000001)

0.19783453580608576


In [298]:
gradient_descent(df6,10000,0.00000001)

(0.07251103749640575,
 array([0.02428294, 0.01902749, 0.02068955, 0.00814059, 0.00886271,
        0.01497106, 0.00979363, 0.03484802, 0.03244248, 0.01487617,
        0.02455769]),
 0.17550495807372465,
 0.1604598318267509)

####  Mean Square error for
#### K_fold Cross validation is 0.17550495807372465

#### Hold out Cross validation is 0.19783453580608576

### (d) Assess the performance of the model.


In [297]:
gradient_descent(df6,10000,0.0001)

(0.5407800555554853,
 array([ 0.06718371, -0.31701839, -0.03782658,  0.05227645, -0.22075078,
         0.06064083, -0.18238542, -0.05836576, -0.09727041,  0.30699806,
         0.35491125]),
 0.016671086676918086,
 0.1291165623302685)

In [255]:
avg_mse=np.mean(mse_list)
print(avg_mse)

0.016905940702166767


In [256]:
np.std(y-y_pred)

0.1291165623302685

### (e) Report search space & the values of the hyperparameters and the parameters of the model.


### hyperparameter 10^-4

In [257]:
gradient_descent(df6,10000,0.0001)

(0.5407800555554853,
 array([ 0.06718371, -0.31701839, -0.03782658,  0.05227645, -0.22075078,
         0.06064083, -0.18238542, -0.05836576, -0.09727041,  0.30699806,
         0.35491125]),
 0.016671086676918086)

### hyperparameter 10^-5

In [258]:
gradient_descent(df6,10000,0.00001)

(0.41317675879494703,
 array([ 0.10875788, -0.24428406,  0.02825643,  0.01284381, -0.09080593,
         0.0433582 , -0.12425074, -0.05104985,  0.03986352,  0.23511143,
         0.37855274]),
 0.017010771138499903)

### hyperparameter 10^-6

In [259]:
gradient_descent(df6,10000,0.000001)

(0.2916034849193102,
 array([ 0.09094878, -0.01483574,  0.10719033,  0.0176734 ,  0.00766509,
         0.03379482, -0.01350934,  0.05449373,  0.121278  ,  0.09527168,
         0.22961534]),
 0.02063684285707668)

### hyperparameter 10^-7

In [264]:
gradient_descent(df6,10000,0.0000001)

(0.26105154460532815,
 array([0.08696452, 0.06115151, 0.07719118, 0.02805255, 0.02974547,
        0.05146651, 0.0308879 , 0.1183063 , 0.11634346, 0.05632919,
        0.09988243]),
 0.025843336522779446)

### hyperparameter 10^-8

In [266]:
gradient_descent(df6,10000,0.00000001)

(0.07251103749640575,
 array([0.02428294, 0.01902749, 0.02068955, 0.00814059, 0.00886271,
        0.01497106, 0.00979363, 0.03484802, 0.03244248, 0.01487617,
        0.02455769]),
 0.17550495807372465)

### (f) Apply batch as well as online optimization algorithms and compare their performance using statistical measures. Compare the time taken by the two algorithms.


## Batch Gradient Descent Alogorithm

In [267]:
gradient_descent(df6,10000,0.0001)

(0.5407800555554853,
 array([ 0.06718371, -0.31701839, -0.03782658,  0.05227645, -0.22075078,
         0.06064083, -0.18238542, -0.05836576, -0.09727041,  0.30699806,
         0.35491125]),
 0.016671086676918086)